### DISCUSSION TITLE
Hilbert Transform Instantaneous Trendline

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#363] [Indicators]
Community discussion for [Hilbert Transform Instantaneous Trendline](https://dotnet.stockindicators.dev/indicators/HtTrendline/#content)

**Comment by Fredrik-C** [comment]
I've done some backtesting on this indicator and I'm very happy to see it being supported by my favorit .Net indicator library.

One thing to keep in mind is that this indicator is way to "slow" to be usefull for buy/sell signals, but is very usefull to be used as part of a screener/stockpicker. The problem with a usefull screening indicator is that you must be able to get a common baseline so that you can benchmark against other stocks, and thats when RSI on HT gets in handy :) 
Hilbert Transform Trendline expressed as a 0-100 range!

**Reply by GordonEldest** [reply]
If I understood your request IMHO
Hilbert Trend deliver a duration, there is no way to make it an indicator as such. 
The duration is a value giving a clue about how long does need a signal (a stocks) to repeat a pattern.
It should be used in // to another indicator to catch the trend. Can be a simple MA, generally SMA (no real reason to use EMA) or RSI using the exact value output by th HT Trend
Then use the output of the indicators.

I personally doesn't see an indicators more suited than another for HT, it depends upon the stocks. 
So I prefer getting the outcome of HT as it is (a duration) then deciding by adding another method to convert it to a score. 
A score you need a ratio. One way can be to make the correlation with an index, then use the duration of HTTrend to make a short correlation base on duration delivered from HT
The ratio of both correlation does give a clue on the current coupling/decoupling. 
Still need to analyze the index trend but generally the less correlated the better (to add to a pool of parameters for screening purpose)

**Comment by GordonEldest** [comment]
My grain of salt. sorry if blunt, no intent,
But aside of perf the interest of "recycling" buffer is also in the discipline it impose to understand what is happening in the process.
It seems that the code here take inspiration from (or it is similar to) Ehlers code. Good and thank you.
Theses codes are largely publicized on the web and available in various specific language of various platforms.
However I like to understand why Ehlers does so many cooking that detrend/smooth/re-smooth. so that finally only 20% of the signal is propagated.
For example I wonder where the "0.0962", "0.5769" coefficients come from. Ehlers doesn't explain it.
If you or anybody else have an explanation I would be please to know it :-)
I guess it as residues of an averaging of a PI/7 rephase applied on a the +&-3 first Hilbert coeff with a lag of 3 (hence [0] to [-6]) but ... unclear I also 'Smell" a 'log' in this because Ehlers works in DeciBels
This opinion because original Hilbert coeff for a small kernel are very simple to compute they are in 1/i i :[1-n] where n is size of kernel

**Reply by DaveSkender** [reply]
John Ehlers likes to apply digital signal processing concepts from Electrical Engineering when analyzing financial market data.  I think what he's trying to do here is to apply the "**Improved Hilbert Transform**(iHT)", a type of electrical frequency high-pass filter; and then perhaps uses smoothing to take off some of the binary squareness.  The "0.0962..." numbers are just an empirical simplification of that iHT formula.  Ehlers didn't invent the **Hilbert Transform**, but is rather just using it here -- there's plenty of scholarly articles out there describing all the glorious advanced mathematics and Fourier transforms behind it.

The best place to get an explanation of Ehlers' methods is from Ehlers himself -- still alive and kickin' with books and training sessions that cover his rational - see [mesasoftware.com](https://mesasoftware.com/).  This particular recipe and full explanation is in his book [Rocket Science for Traders](http://www.amazon.com/Rocket-Science-Traders-Processing-Applications/dp/0471405671/ref=sr_1_3?ie=UTF8&qid=1419372320&sr=8-3&keywords=John+Ehlers) -- a highly recommended purchase.

Ehlers also uses Hilbert Transform in his [MAMA](https://daveskender.github.io/Stock.Indicators/indicators/Mama/#content) indicator, among others.

**Reply by DaveSkender** [reply]
Smoothing is not uncommon in indicator recipes overall.  In fact, probably 1/3 to 1/2 of the indicators in this library uses some form of smoothing to reduce noise and to help traders avoid [whipsaw](https://www.investopedia.com/terms/w/whipsaw.asp).
